contrib/zenith/pagestore_smgr.c

```diff
@@ -412,6 +434,8 @@ zenith_get_request_lsn(bool nonrel)
 			elog(DEBUG1, "zenith_get_request_lsn GetFlushRecPtr lsn %X/%X",
 				 (uint32) ((lsn) >> 32), (uint32) (lsn));
 		}
+		else
+			lsn = zm_adjust_lsn(lsn);
```

Comment:
Hmm, I think this is too late to be doing this. It means that the "last-written-page LSN" can be ahead of the last flush LSN, but I think we should prevent that from happening in the first place. We should never write out a page with LSN that hasn't been flushed yet, in fact maybe we should add an assertion for that in SetLastWrittenPageLSN().

Reply:
According to the log this situation (last-written-page-lsn > flush-lsn) happens quite frequently. It also seems to be strange to me because it violates the main principle of WAL logging: page can not be written to the database until it is flushed in log. But looks like Postgres violates this rule at least for new pages. I just insert asserts and immediately find out violation of this rule: _bt_blwritepage is wal recording page using log_newpage (nbtstort.c:646), which assigns Page LSN but doesn't perform  flush. Then we write this page nbtsort.c:683 using smgrwrite and here pageLSN is greater than GetFlushRecPtr(). Looks like it should not cause possible database corruption because it happens only while index creation and if it is interrupted by crash, then index  relation should be completely removed and we should not try to traverse may be incorrectly written pages of this index.

Below is stacktrace:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007f45318b28b1 in __GI_abort () at abort.c:79
#2  0x000055f1d788bc2c in ExceptionalCondition (conditionName=0x7f453166f924 "lsn <= flushLsn", errorType=0x7f453166f762 "FailedAssertion", 
    fileName=0x7f453166f913 "pagestore_smgr.c", lineNumber=243) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/utils/error/assert.c:69
#3  0x00007f453166e16c in zenith_wallog_page (reln=0x55f1d8a69460, forknum=MAIN_FORKNUM, blocknum=0, buffer=0x55f1d8ae0100 "") at pagestore_smgr.c:243
#4  0x00007f453166ef6f in zenith_write (reln=0x55f1d8a69460, forknum=MAIN_FORKNUM, blocknum=0, buffer=0x55f1d8ae0100 "", skipFsync=true) at pagestore_smgr.c:825
#5  0x000055f1d76d6377 in smgrwrite (reln=0x55f1d8a69460, forknum=MAIN_FORKNUM, blocknum=0, buffer=0x55f1d8ae0100 "", skipFsync=true)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/storage/smgr/smgr.c:519
#6  0x000055f1d725d643 in _bt_blwritepage (wstate=0x7ffc6f1f6a40, page=0x55f1d8ae0100 "", blkno=0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/nbtree/nbtsort.c:683
#7  0x000055f1d725e62e in _bt_uppershutdown (wstate=0x7ffc6f1f6a40, state=0x55f1d8abf588)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/nbtree/nbtsort.c:1172
#8  0x000055f1d725f38a in _bt_load (wstate=0x7ffc6f1f6a40, btspool=0x55f1d8abed08, btspool2=0x0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/nbtree/nbtsort.c:1418
#9  0x000055f1d725d343 in _bt_leafbuild (btspool=0x55f1d8abed08, btspool2=0x0)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/nbtree/nbtsort.c:574
#10 0x000055f1d725cdc0 in btbuild (heap=0x7f453276bd20, index=0x7f45327742d8, indexInfo=0x55f1d8ad44c8)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/access/nbtree/nbtsort.c:332
#11 0x000055f1d72f0fc1 in index_build (heapRelation=0x7f453276bd20, indexRelation=0x7f45327742d8, indexInfo=0x55f1d8ad44c8, isreindex=false, parallel=true)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/catalog/index.c:2971
#12 0x000055f1d72eddae in index_create (heapRelation=0x7f453276bd20, indexRelationName=0x55f1d8ad4688 "pgbench_branches_pkey", indexRelationId=16400, parentIndexRelid=0, 
    parentConstraintId=0, relFileNode=0, indexInfo=0x55f1d8ad44c8, indexColNames=0x55f1d8ad4630, accessMethodObjectId=403, tableSpaceId=0, 
    collationObjectId=0x55f1d8ad52f0, classObjectId=0x55f1d8ad5310, coloptions=0x55f1d8ad5330, reloptions=0, flags=3, constr_flags=0, allow_system_table_mods=false, 
    is_internal=false, constraintId=0x7ffc6f1f6dc4) at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/catalog/index.c:1229
#13 0x000055f1d73e8b20 in DefineIndex (relationId=16393, stmt=0x55f1d8abade0, indexRelationId=0, parentIndexId=0, parentConstraintId=0, is_alter_table=true, 
    check_rights=true, check_not_in_use=false, skip_build=false, quiet=false)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/commands/indexcmds.c:1137
#14 0x000055f1d741db4c in ATExecAddIndex (tab=0x55f1d89fb820, rel=0x7f453276bd20, stmt=0x55f1d8abade0, is_rebuild=false, lockmode=8)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/commands/tablecmds.c:8411
#15 0x000055f1d7415a64 in ATExecCmd (wqueue=0x7ffc6f1f7178, tab=0x55f1d89fb820, cmd=0x55f1d8abb590, lockmode=8, cur_pass=8, context=0x7ffc6f1f7370)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/commands/tablecmds.c:4911
#16 0x000055f1d741532b in ATRewriteCatalogs (wqueue=0x7ffc6f1f7178, lockmode=8, context=0x7ffc6f1f7370)
    at /home/knizhnik/zenith.main/tmp_install/build/../../vendor/postgres/src/backend/commands/tablecmds.c:4787
```